From: Stanislav Malyshev <stas@php.net>
Date: Mon Sep 28 17:12:35 2015 -0700
Subject: [PATCH] FIx bug #70433 - Uninitialized pointer in phar_make_dirstream when zip entry filename is "/"

5.4 backport: https://github.com/shivammathur/php-src-backports/commit/16f6c66b816e3a506f3c5ddb812e1557f8742593
Binary test fixtures not included: ext/phar/tests/bug70433.zip
---
diff --git a/ext/phar/tests/bug70433.phpt b/ext/phar/tests/bug70433.phpt
new file mode 100644
--- /dev/null
+++ b/ext/phar/tests/bug70433.phpt
@@ -0,0 +1,23 @@
+--TEST--
+Phar - bug #70433 - Uninitialized pointer in phar_make_dirstream when zip entry filename is "/"
+--SKIPIF--
+<?php if (!extension_loaded("phar") || !extension_loaded("zlib")) die("skip"); ?>
+--FILE--
+<?php
+$phar = new PharData(__DIR__."/bug70433.zip");
+var_dump($phar);
+$meta = $phar->getMetadata();
+var_dump($meta);
+?>
+DONE
+--EXPECTF--
+object(PharData)#1 (3) {
+  ["pathName":"SplFileInfo":private]=>
+  string(0) ""
+  ["glob":"DirectoryIterator":private]=>
+  bool(false)
+  ["subPathName":"RecursiveDirectoryIterator":private]=>
+  string(0) ""
+}
+NULL
+DONE
diff --git a/ext/phar/util.c b/ext/phar/util.c
--- a/ext/phar/util.c
+++ b/ext/phar/util.c
@@ -2248,7 +2248,7 @@ void phar_add_virtual_dirs(phar_archive_data *phar, char *filename, int filename
 
 	while ((s = zend_memrchr(filename, '/', filename_len))) {
 		filename_len = s - filename;
-		if (FAILURE == zend_hash_add_empty_element(&phar->virtual_dirs, filename, filename_len)) {
+		if (!filename_len || FAILURE == zend_hash_add_empty_element(&phar->virtual_dirs, filename, filename_len)) {
 			break;
 		}
 	}
diff --git a/ext/phar/zip.c b/ext/phar/zip.c
--- a/ext/phar/zip.c
+++ b/ext/phar/zip.c
@@ -396,7 +396,9 @@ foundit:
 
 		if (entry.filename[entry.filename_len - 1] == '/') {
 			entry.is_dir = 1;
-			entry.filename_len--;
+			if(entry.filename_len > 1) {
+				entry.filename_len--;
+			}
 			entry.flags |= PHAR_ENT_PERM_DEF_DIR;
 		} else {
 			entry.is_dir = 0;
